src/ucs/memory/memtype_cache.c

```diff
@@ -31,6 +31,34 @@ typedef enum {
 } ucs_memtype_cache_action_t;
 
 
+static UCS_CLASS_INIT_FUNC(ucs_memtype_cache_t);
+UCS_CLASS_DECLARE_NAMED_NEW_FUNC(ucs_memtype_cache_create, ucs_memtype_cache_t);
+UCS_CLASS_DECLARE_NAMED_DELETE_FUNC(ucs_memtype_cache_destroy,
+                                    ucs_memtype_cache_t);
```

Comment:
```diff
diff --git a/src/ucs/memory/memtype_cache.c b/src/ucs/memory/memtype_cache.c
index 7d2487dd9..08e245f5d 100644
--- a/src/ucs/memory/memtype_cache.c
+++ b/src/ucs/memory/memtype_cache.c
@@ -34,9 +34,9 @@ typedef enum {
 
 
 static UCS_CLASS_INIT_FUNC(ucs_memtype_cache_t);
-UCS_CLASS_DECLARE_NAMED_NEW_FUNC(ucs_memtype_cache_create, ucs_memtype_cache_t);
-UCS_CLASS_DECLARE_NAMED_DELETE_FUNC(ucs_memtype_cache_destroy,
-                                    ucs_memtype_cache_t);
+//UCS_CLASS_DECLARE_NAMED_NEW_FUNC(ucs_memtype_cache_create, ucs_memtype_cache_t);
+//UCS_CLASS_DECLARE_NAMED_DELETE_FUNC(ucs_memtype_cache_destroy,
+//                                    ucs_memtype_cache_t);
 
 
 static UCS_F_ALWAYS_INLINE ucs_memtype_cache_t *ucs_memtype_cache_get_global()
```
```
  CC       memory/libucs_la-memtype_cache.lo
  CC       sys/libucs_la-init.lo
../../../src/ucs/memory/memtype_cache.c: In function ‘ucs_memtype_cache_get_global’:
../../../src/ucs/memory/memtype_cache.c:53:18: error: implicit declaration of function ‘ucs_memtype_cache_create’; did you mean ‘ucs_me
mtype_cache_update’? [-Werror=implicit-function-declaration]
         status = ucs_memtype_cache_create(&memtype_cache);
                  ^~~~~~~~~~~~~~~~~~~~~~~~
                  ucs_memtype_cache_update
../../../src/ucs/memory/memtype_cache.c:53:18: error: nested extern declaration of ‘ucs_memtype_cache_create’ [-Werror=nested-externs]
In file included from ../../../src/ucs/memory/memtype_cache.c:14:0:
../../../src/ucs/memory/memtype_cache.c: At top level:
../../../src/ucs/memory/memtype_cache.c:394:33: error: conflicting types for ‘ucs_memtype_cache_create’
 UCS_CLASS_DEFINE_NAMED_NEW_FUNC(ucs_memtype_cache_create, ucs_memtype_cache_t,
                                 ^
/gpfs/fs1/akvenkatesh/ucx-github/build-own/../src/ucs/type/class.h:217:18: note: in definition of macro ‘UCS_CLASS_DECLARE_NAMED_NEW_FU
NC’
     ucs_status_t _name(UCS_PP_FOREACH(_UCS_CLASS_INIT_ARG_DEFINE, _, \
                  ^~~~~
../../../src/ucs/memory/memtype_cache.c:394:1: note: in expansion of macro ‘UCS_CLASS_DEFINE_NAMED_NEW_FUNC’
 UCS_CLASS_DEFINE_NAMED_NEW_FUNC(ucs_memtype_cache_create, ucs_memtype_cache_t,
 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../../src/ucs/memory/memtype_cache.c:53:18: note: previous implicit declaration of ‘ucs_memtype_cache_create’ was here
         status = ucs_memtype_cache_create(&memtype_cache);
                  ^~~~~~~~~~~~~~~~~~~~~~~~
```

Reply:
like this, since we don't need to expose memtype_create/destroy now:
```diff
diff --git a/src/ucs/memory/memtype_cache.c b/src/ucs/memory/memtype_cache.c
index 7d2487dd94..4fa37377e2 100644
--- a/src/ucs/memory/memtype_cache.c
+++ b/src/ucs/memory/memtype_cache.c
@@ -32,12 +32,10 @@ typedef enum {
     UCS_MEMTYPE_CACHE_ACTION_REMOVE
 } ucs_memtype_cache_action_t;
 
-
 static UCS_CLASS_INIT_FUNC(ucs_memtype_cache_t);
-UCS_CLASS_DECLARE_NAMED_NEW_FUNC(ucs_memtype_cache_create, ucs_memtype_cache_t);
-UCS_CLASS_DECLARE_NAMED_DELETE_FUNC(ucs_memtype_cache_destroy,
-                                    ucs_memtype_cache_t);
+static UCS_CLASS_CLEANUP_FUNC(ucs_memtype_cache_t);
 
+UCS_CLASS_DEFINE(ucs_memtype_cache_t, void);
 
 static UCS_F_ALWAYS_INLINE ucs_memtype_cache_t *ucs_memtype_cache_get_global()
 {
@@ -50,7 +48,7 @@ static UCS_F_ALWAYS_INLINE ucs_memtype_cache_t *ucs_memtype_cache_get_global()
 
     if (ucs_memtype_cache_global_instance == NULL) {
         ucs_spin_lock(&ucs_memtype_cache_global_instance_lock);
-        status = ucs_memtype_cache_create(&memtype_cache);
+        status = UCS_CLASS_NEW(ucs_memtype_cache_t, &memtype_cache);
         if (status != UCS_OK) {
             ucs_spin_unlock(&ucs_memtype_cache_global_instance_lock);
             return NULL;
@@ -390,16 +388,10 @@ static UCS_CLASS_CLEANUP_FUNC(ucs_memtype_cache_t)
     pthread_rwlock_destroy(&self->lock);
 }
 
-UCS_CLASS_DEFINE(ucs_memtype_cache_t, void);
-UCS_CLASS_DEFINE_NAMED_NEW_FUNC(ucs_memtype_cache_create, ucs_memtype_cache_t,
-                                ucs_memtype_cache_t)
-UCS_CLASS_DEFINE_NAMED_DELETE_FUNC(ucs_memtype_cache_destroy, ucs_memtype_cache_t,
-                                   ucs_memtype_cache_t)
-
 void ucs_memtype_cache_global_cleanup()
 {
     if (ucs_memtype_cache_global_instance) {
-        ucs_memtype_cache_destroy(ucs_memtype_cache_global_instance);
+        UCS_CLASS_DELETE(ucs_memtype_cache_t, ucs_memtype_cache_global_instance);
     }
 }
 ```

Reply:
thanks!